### PR TITLE
vmap proof of concept

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#32558 vmap proof of concept**
* #32557 Add missing C10_API to dispatch key TLS setter/getters
* #32533 Improve documentation in dispatcher; remove unnecessary optional

This very simple proof of concept shows how dispatcher "modes" can be used
to implement JAX style vmap (see https://jax.readthedocs.io/en/latest/jax.html#vectorization-vmap)
When batching is turned on, operations route to "batched" versions which assume that
the first dimension is a batch dimension.

For simplicity, the proof of concept only implements the batched version of
torch.mv, but it's easy to see how it could be extended for more operations.
The PoC also assumes that ALL tensors are batched (a more sophisticated
implementation would track this on a per tensor basis, so we can insert
broadcasts as necessary.)  I wanted to show per-example gradients in this
PoC but some more invasive autograd changes are ncessary.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

### CODE REVIEW

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit b9d6ac58:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=b9d6ac58f6c8746377e115866023cfb6da580e33).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4327833) pytorch_linux_xenial_py2_7_9_build (1/1)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=90081917) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=90081917))
<details>
<summary>
<code>Jan 23 23:50:21 /var/lib/jenkins/workspace/c10/util/Exception.h:154:83: error: throw will always call terminate() [-Werror=terminate]</code>
</summary>

```
Jan 23 23:50:12 [ 85%] Building CXX object caffe2/CMakeFiles/kernel_lambda_test.dir/__/aten/src/ATen/core/boxing/kernel_lambda_test.cpp.o 
Jan 23 23:50:13 [ 85%] Linking CXX executable ../bin/string_ops_test 
Jan 23 23:50:13 [ 85%] Built target string_ops_test 
Jan 23 23:50:13 Scanning dependencies of target apply_utils_test 
Jan 23 23:50:13 [ 85%] Building CXX object caffe2/CMakeFiles/apply_utils_test.dir/__/aten/src/ATen/test/apply_utils_test.cpp.o 
Jan 23 23:50:21 In file included from /var/lib/jenkins/workspace/c10/core/Device.h:5:0, 
Jan 23 23:50:21                  from /var/lib/jenkins/workspace/c10/core/Allocator.h:6, 
Jan 23 23:50:21                  from /var/lib/jenkins/workspace/aten/src/ATen/ATen.h:3, 
Jan 23 23:50:21                  from /var/lib/jenkins/workspace/aten/src/ATen/test/vmap_test.cpp:3: 
Jan 23 23:50:21 /var/lib/jenkins/workspace/aten/src/ATen/test/vmap_test.cpp: In destructor '{anonymous}::VMapMode::~VMapMode()': 
Jan 23 23:50:21 /var/lib/jenkins/workspace/c10/util/Exception.h:154:83: error: throw will always call terminate() [-Werror=terminate] 
Jan 23 23:50:21    throw ::c10::err_type({__func__, __FILE__, static_cast<uint32_t>(__LINE__)}, msg) 
Jan 23 23:50:21                                                                                    ^ 
Jan 23 23:50:21 /var/lib/jenkins/workspace/c10/util/Exception.h:216:5: note: in expansion of macro 'C10_THROW_ERROR' 
Jan 23 23:50:21      C10_THROW_ERROR(Error, ::c10::str(        \ 
Jan 23 23:50:21      ^~~~~~~~~~~~~~~ 
Jan 23 23:50:21 /var/lib/jenkins/workspace/aten/src/ATen/test/vmap_test.cpp:30:5: note: in expansion of macro 'TORCH_INTERNAL_ASSERT' 
Jan 23 23:50:21      TORCH_INTERNAL_ASSERT(vmap_level > 0); 
Jan 23 23:50:21      ^~~~~~~~~~~~~~~~~~~~~ 
Jan 23 23:50:21 /var/lib/jenkins/workspace/c10/util/Exception.h:154:83: note: in C++11 destructors default to noexcept 
Jan 23 23:50:21    throw ::c10::err_type({__func__, __FILE__, static_cast<uint32_t>(__LINE__)}, msg) 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

**Comment by ezyang** [comment]
cc'ing @yaroslavvb @phizaz who have discussed related problems on per-sample gradient at https://discuss.pytorch.org/t/per-sample-gradient-should-we-design-each-layer-differently/57293/15

This trick doesn't solve the problem of having to reimplement kernels in some cases, but it does suggest a potential API here.
